docs/cn_omf.md

```diff
@@ -203,16 +214,92 @@ This is followed by the bloom filter.
+stores its hlog bytes and the kvset's hlog can be composed using the individual
+kblock hlogs.
 
 ## VBlocks (Value blocks)
 
-VBlocks consist of a header followed by all the values.  The offset of a specifc value and its length are obtained from the kmd entry for this value which is linked to the corresponding key.
-
-### VBlock header
+VBlocks consist of a header followed by all the values. The offset of a specifc
+value and its length are obtained from the kmd entry for this value which is
+linked to the corresponding key.
+
+### VBlock trailer
+
+    +---------------------+
+    | u32 vbh_magic       |
+    | u32 vbh_version     |
+    | u64 vbh_vgroup      |
+    | u32 vbh_min_key_off |
+    | u32 vbh_min_key_len |
+    | u32 vbh_max_key_off |
+    | u32 vbh_max_key_len |
+    |---------------------|
+    | min key             | 1344 bytes
+    |---------------------|
+    | max key             | 1344 bytes
+    +---------------------+
+
+## Kvsets (Key-value sets)
+
+Kvsets are composed of kblocks and vblocks which are related to each other.
+
+### Kvset trailer
+
+Mblock consisting of kvset metadata.
+
+    +----------+--------+--------+-------+----------+
+    |          |        |        |       |          |
+    | Metadata | Kblock | Vblock | PTree | Hyperlog |
+    |  Header  |  Trim  | Index  |       |          |
+    |          |        | Adjust |       |          |
+    +----------+--------+--------+-------+----------+
+
+#### Metadata header
+
+    +---------------------------+
+    | u32 kmh_num_kblocks       |
+    | u32 kmh_num_vblocks       |
+    | u32 kmh_num_vgroups       |
+    | u32 kmh_kblk_trim_off     |
+    | u32 kmh_kblk_trim_len     |
+    | u32 kmh_vgroup_layout_off |
+    | u32 kmh_vgroup_layour_len |
+    | u32 kmh_vblk_idx_adj_off  |
+    | u32 kmh_vblk_idx_adj_len  |
+    | u32 kmh_ptree_off         |
+    | u32 kmh_ptree_len         |
+    | u32 kmh_hlog_off          |
+    | u32 kmh_hlog_len          |
+    +---------------------------+
+
+#### Kblock trim
+
+Kblock trim masks parts of a kblock's data. During a kvset split, a portion of
+a kblock will be hidden. Each segment is the trim associated with 1 kblock.
+
+    +------------------+
+    | ...              |
+    +------------------+
+    | u16 kt_start_pg  |
+    | u8  kt_start_off |
+    | u16 kt_end_pg    |
+    | u8  kt_end_off   |
+    +------------------+
+    | ...              |
+    +------------------+
+
+#### Vblock index adjust
+
+The index indicates that from the previous index + 1 to the current index, all
+index adjustments must be subtracted by the accompanying adjustment value.
 
     +-------------------+
-    | u32 vbh_magic     |
-    | u32 vbh_version   |
-    | u64 vbh_vgroup    |
+    | ...               |
+    +-------------------+
+    | u32 vbia_vblk_idx |
+    | u32 vbia_adj      |
+    +-------------------+
+    | ...               |
     +-------------------+
```

Comment:
What's the max possible value for vblock index in the KMD?

Reply:
API takes a uint and calls `encode_hg16_32k()` on it:

```
/* HG16_32K encoding
 *
 *   Max Value    Encoding
 *   127          0xxxxxxx
 *   32K-1        1xxxxxxx xxxxxxxx
 */
#define HG16_32K_MAX 0x7fffu
```

So I should use a `uint16_t` here it looks like?

Reply:
Comment is no extremely clear, although I assume this is some well-known encoding, so I can just read up on it.

Reply:
> So I should use a uint32_t here it looks like?
Yes.

Reply:
The encoding uses less space for small values, but the in-memory version must be able to handle the largest value.  In this case a 16-bit uint suffices.